### DISCUSSION TITLE
Finding Used Ports on Mirth Connect Server

### DISCUSSION THREAD

**Discussion by JeffBenedict** [nextgenhealthcare/connect#5250] [General]
I'm sure this has been asked a thousand times, but searching for "Open Ports", "Free Ports", "Unused Ports", etc, I can't seem to find an answer.  What method do folks use to find the inbound ports that have been used/allocated in their Mirth Connect installs so that they can keep track of what's available and what can be used.  Thanks in advance.

**Comment by pacmano1** [comment]
It is a process thing here including adding a tag with the port number to the channel to make it easier to see.   There isn't a great "easy button" for this problem, you can see in issues recent requests to add this functionality as standard.   There are options of course including sourcing stuff from DBs, etc.

**Comment by aungktun** [comment]
What port is being used can definitely be tracked by pulling channel meta
out of db table.

**Reply by pacmano1** [reply]
Not always.  Ports can be set with a var.

**Comment by aungktun** [comment]
Yes, true

**Comment by jonbartels** [comment]
In the paid space there are a few options. Both Zen and NextGen have channel analysis tools which will report on ports used by source and destination connectors. Zen has (had?) a tool as part of their Insight offering that would also do this analysis on your MC instance. NextGen has you upload a server export to their Success Community portal. It wouldn't surprise me if Innovar was working on something but I don't know their offerings.

Those tools can't be doing anything too magical. Fundamentally they would have to look at the channel definition, inspect the source connector, then for the connector types that allow ports (network connectors as opposed to say a file reader) the port field can be parsed out. This can be done either in the MC database in the `channel` table, by looking at each channel export individually, or a global server export. I vaguely remember someone having a DB query that did this but I cannot find the project or gist.

@aungktun  did some work on this at https://github.com/nextgenhealthcare/connect/discussions/5245 . His query does not cover all the connector types but is a great start on the problem. You can see how he extracts connector details from the MC channel XML. I think its a great baseline, the next steps are just to add different connector types.

Some additional ideas around adding this to the dashboard are:
* https://github.com/nextgenhealthcare/connect/issues/5152
* I wrote https://github.com/jonbartels/mirth-connect-dashboard-count in response to the above discussion, note this plugin is affected by a bug in MC. It can only show ports on the dashboard and not on the channel list.
* https://github.com/nextgenhealthcare/connect/issues/2115 is OOOOLD but was recently reopened

I hope these references help you find a starting point. As you build on others work or write your own code, please do share that with the community. You're not the only one pecking at the problem!

**Comment by aungktun** [comment]
Thank you Jon for compiling all info in one place.
